test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c

```diff
@@ -329,22 +330,47 @@ VMInit(jvmtiEnv *jvmti, JNIEnv* jnii, jthread thread) {
     }
 }
 
+static void JNICALL
+VMDeath(jvmtiEnv *jvmti, JNIEnv* jni) {
+    jvmtiError err;
+
+    // Block ClassPrepare events while this callback is executed.
+    err  = (*jvmti)->RawMonitorEnter(jvmti, event_mon);
+    check_jvmti_error(jvmti, "VMDeath event: Failed in RawMonitorEnter", err);
+
+    printf("VMDeath event\n");
+
+    err = (*jvmti)->RawMonitorExit(jvmti, event_mon);
+    check_jvmti_error(jvmti, "VMDeath event: Failed in RawMonitorExit", err);
+}
+
 static void JNICALL
 ClassPrepare(jvmtiEnv *jvmti, JNIEnv *env, jthread thread, jclass klass) {
     static const jint EVENTS_LIMIT = 2;
     static       jint event_no = 0;
-    jthread cur_thread = get_cur_thread(jvmti);
     jvmtiPhase phase;
     intptr_t exp_val = 777;
     intptr_t act_val;
+    jvmtiError err;
+
+    // Block VMDeath event and other ClassPrepare events while this callback is executed.
+    // Sync with VMDeath event guards against JVMTI_ERROR WRONG_PHASE.
+    err = (*jvmti)->RawMonitorEnter(jvmti, event_mon);
+    check_jvmti_error(jvmti, "ClassPrepare event: Failed in RawMonitorEnter", err);
 
     get_phase(jvmti, &phase);
     if (phase != JVMTI_PHASE_START && phase != JVMTI_PHASE_LIVE) {
-        printf("  ## Error: unexpected phase: %d, expected: %d or %d\n",
-               phase, JVMTI_PHASE_START, JVMTI_PHASE_LIVE);
+        if (phase != JVMTI_PHASE_DEAD) {
```

Comment:
After this check is made, what prevents JVMTI from moving to the JVMTI_PHASE_DEAD before the get_cur_thread() call below? I'm not sure if the phase is changed to JVMTI_PHASE_DEAD before or after calling VMDeath, but it seems in either case you can end up with the phase changing after making this check. There is nothing that stops the phase from changing while holding the raw monitor.

Reply:
Thanks, Chris.
The `JvmtiExport::post_vm_death()` posts the `VMDeath` before changing the phase to `JVMTI_PHASE_DEAD`.
The `VMDeath` callback is blocked on the `RawMonitor` while it is grabbed by the `ClassPrepare` event callback.
But you are right, there is a sync gap in `JvmtiExport::post_vm_death()` between the `VMDeath` callback execution and phase changing. So, I'll go with your suggestion and introduce a flag.

Reply:
Added `is_vm_dead` flag.